agent/testflinger_agent/agent.py

```diff
@@ -15,14 +15,31 @@
 import json
 import logging
 import os
+from pathlib import Path
 import shutil
+import tarfile
+import tempfile
 
 from testflinger_agent.job import TestflingerJob
 from testflinger_agent.errors import TFServerError
+from testflinger_agent.config import ATTACHMENTS_DIR
 
 logger = logging.getLogger(__name__)
 
 
+def secure_filter(member, path):
+    """Combine the `data` filter with custom attachment filtering
+
+    Makes sure that the starting folder for all attachments coincides
+    with one of the supported phases, i.e. that the attachment archive
+    has been created properly and no attachment will be extracted to an
+    unexpected location.
+    """
+    if not member.name.startswith(("provision", "firmware_update", "test")):
+        raise tarfile.OutsideDestinationError(member, path)
```

Comment:
1. What if my file is called `provisionfoo/bar`?
2. What if my file is called `provision/../foo`?

The [`data_filter()`](https://docs.python.org/3/library/tarfile.html#tarfile.data_filter) and [`tar_filter()`](https://docs.python.org/3/library/tarfile.html#tarfile.tar_filter) functions don't seem to block that (tar_filter is mostly concerned with absolute paths, and data_filter is mostly concerned with links and special files).

Something like [`os.path.normpath()`](https://docs.python.org/3/library/os.path.html#os.path.normpath) (to resolve case 2 above) and checking for `.startswith("something/")` (to resolve case 1 above) might be a first step, haven't thought about whether that's good enough).